Concurrent form requests can fail (XML parser is not re-entrant)

If I make a series of requests for forms concurrently, e.g. lots of HTTP GET requests for `http://ping.buendia.org:9000/openmrs/ws/rest/v1/projectbuendia/xforms/buendia-form-patient_status?v=full`, some of them will fail.

This happens because `XmlUtil` is written so that _all requests and threads_ use a single common `DocumentBuilder` object, which cannot parse multiple documents at the same time.

The exception thrown looks like this:

```
Caused by: org.xml.sax.SAXException: FWK005 parse may not be called while parsing.
	at org.apache.xerces.parsers.DOMParser.parse(Unknown Source)
	at org.apache.xerces.jaxp.DocumentBuilderImpl.parse(Unknown Source)
	at org.openmrs.projectbuendia.webservices.rest.XmlUtil.parse(XmlUtil.java:135)
	at org.openmrs.projectbuendia.webservices.rest.XformResource.convertToOdkCollect(XformResource.java:148)
```

or this:

```
Caused by: java.lang.NullPointerException
	at org.apache.xerces.parsers.AbstractDOMParser.startDocument(Unknown Source)
	at org.apache.xerces.impl.dtd.XMLDTDValidator.startDocument(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentScannerImpl.startEntity(Unknown Source)
	at org.apache.xerces.impl.XMLVersionDetector.startDocumentParsing(Unknown Source)
```

--- a/openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/XformResource.java
+++ b/openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/XformResource.java
@@ -146,7 +146,7 @@ public class XformResource extends AbstractReadOnlyResource<Form> {
         // seems
         // to do odd things... so instead, we import it into a new document.
         Document oldDoc = XmlUtil.parse(xml);
-        Document doc = XmlUtil.getDocumentBuilder().newDocument();
+        Document doc = XmlUtil.createDocumentBuilder().newDocument();
         Element root = (Element) doc.importNode(oldDoc.getDocumentElement(), true);
         root = (Element) doc.renameNode(root, HTML_NAMESPACE, "h:form");
         doc.appendChild(root);

--- a/openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/XmlUtil.java
+++ b/openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/XmlUtil.java
@@ -30,19 +30,6 @@ import javax.xml.parsers.ParserConfigurationException;
 
 /** XML manipulation functions. */
 public class XmlUtil {
-    private static final DocumentBuilder documentBuilder;
-
-    static {
-        try {
-            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
-            factory.setNamespaceAware(true);
-            factory.setIgnoringComments(true);
-            documentBuilder = factory.newDocumentBuilder();
-        } catch (ParserConfigurationException e) {
-            throw new RuntimeException(e);
-        }
-    }
-
     /** Converts a NodeList to an Iterable of Elements. */
     public static Iterable<Element> toElementIterable(NodeList nodeList) {
         List<Element> elements = new ArrayList<>(nodeList.getLength());
@@ -125,13 +112,23 @@ public class XmlUtil {
         return ret;
     }
 
-    /** Returns a namespace-aware DocumentBuilder. */
-    public static DocumentBuilder getDocumentBuilder() {
-        return documentBuilder;
+    /** Constructs a new namespace-aware DocumentBuilder. */
+    public static DocumentBuilder createDocumentBuilder() {
+        try {
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+            factory.setNamespaceAware(true);
+            factory.setIgnoringComments(true);
+            return factory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     /** Parses the given XML string to produce a Document. */
     public static Document parse(String xml) throws SAXException, IOException {
-        return documentBuilder.parse(new InputSource(new StringReader(xml)));
+        // DocumentBuilder is not thread-safe; parsing multiple documents concurrently
+        // will cause the error "FWK005 parse may not be called while parsing".
+        // So, we construct a new DocumentBuilder every time we parse something.
+        return createDocumentBuilder().parse(new InputSource(new StringReader(xml)));
     }
 }

